Automattic/wp-calypso
# Move Gutenberg opt-out checks to the server

Previous art: https://github.com/Automattic/wp-calypso/pull/32735/files

Paired with D33359-code, this contains improvements to how we denied the classic editor to FSE and sites with template-first themes.

For some context, the API response to `/sites/:site/gutenberg` includes `editor_mobile`, `editor_web`, and and `opt_{out/in}`. If the editor is Classic, then we would see would be `opt_in` and if the editor is Gutenberg, then we see `opt_out`.

Before this PR, all `opt_out` responses were ignored. In #35119, @mmtr added support for disabling the opt-in dialogue. If `opt_in` is false and the current editor is classic, then the user is not able to see dialogues for opting in.

With this PR, `opt_out` responses are rolled into the same bit of state. This works since the state is not used to store the opt-in URL, but used to verify if the user is allowed to see opt-x dialogues. _I know the naming here is confusing, but I feel I am not necessarily introducing the confusion. If needed, I can move this to a separate bit of state, but that would be enough extra stuff that I felt it wasn't necessary._ 

Additionally, the check for `showOptOut` included `isGutenbergOptInEnabled( state, siteId );`, which got updated to change behavior in the above PR. This PR also makes that behavior consistent.

### Changes proposed in this Pull Request
* If the `opt_out` field is false in the API and the current editor is not classic (i.e. it is one of the Gutenberg editor), then the user is not allowed to see the "switch to classic" dialogue.
* Remove client-side checks for FSE and for SPT themes. These checks are now centralized on the server in D33359-code.

### Testing instructions
1. Apply D33359-code to your sandbox (and make sure you `svn up`).
2. Sandbox the public API
3. Try each of the test scenarios below

**Test 1**
Sandbox a site with the following:
- Not FSE
- Not using a template first theme
- Your editor is set to Gutenberg

Edit a page and make sure you **can** see the opt-out options in these places:
<img width="561" alt="Screen Shot 2019-10-01 at 8 16 57 PM" src="https://user-images.githubusercontent.com/6265975/66017253-e16cbe80-e48e-11e9-92ee-41a780084e2c.png">

**Test 2**
Sandbox a site with the following:
- Has FSE enabled
- Your editor is set to Gutenberg

Edit a page and make sure you **can not** see the opt-out options in those places:
<img width="521" alt="Screen Shot 2019-10-01 at 8 26 41 PM" src="https://user-images.githubusercontent.com/6265975/66017291-03fed780-e48f-11e9-8e1f-6268ada4a25f.png">

**Test 3**
Sandbox a site with the following:
- Does not have FSE
- Theme is twenty nineteen
- Your editor is set to Gutenberg

Edit a page and make sure you **can not** see the opt-out options (see above screenshot).

**Test 4**
Sandbox a site with the following:
- Does not have FSE
- Your editor is set to Classic

Verify existing behavior. Edit a page and make sure you **can** see opt-in dialogues:
<img width="655" alt="Screen Shot 2019-10-01 at 9 07 52 PM" src="https://user-images.githubusercontent.com/6265975/66017412-9606e000-e48f-11e9-86e9-a8a2474253fe.png">

Fixes #36490
